### ISSUE TITLE
Consider adding kwargs to chop

### ISSUE THREAD

**Issue by bkamins** [JuliaLang/julia#37397]
Very often (~ 1/week) I get questions how to do things like `str[5:8]`, but using character indexing not byte indexing.

The way to do it is currently `chop(str, head=4, tail=length(str)-8)`.

This is a bit uncomfortable and requires using `length` which: a) is expensive, b) does not compose well with broadcasting, as you have to create an anonymous function like this `(str -> chop(str, head=4, tail=length(str)-8)).(str_vec)`.

The question is if it would be reasonable to add two kwargs:
* `stop` which would chop `length(str)-stop` characters from end (and if passed then `tail` is not allowed to be passed)
* `start` which would chop `start-1` characters from start (and if passed then `head` is not allowed to be passed)

Then you could just write `chop(str, start=5, stop=8)` instead.

Adding `start` is less crucial (as it is just off-by-one with `head`), but adding `stop` would solve the problems I have described above.

Of course we could use another function for this, but I propose to change `chop` as then we do not add a new name to the list of exported names from Base.

CC @StefanKarpinski

**Comment by StefanKarpinski** [comment]
This could be good. Another unfortunate thing about the current API is that if you do `chop(s, head=3)` it _also_ chops off the last character, which is pretty awful. I really doubt that anyone is relying on that (since it's totally ridiculous), so we can probably assume that anyone using `chop` to remove characters from the front is calling it as `chop(s, head=3, tail=0)`, so arguably we could fix that. Here's a potential implementation:
```jl
function chop(
    str::AbstractString;
    head::Integer = nothing,
    tail::Integer = nothing,
    start::Integer = nothing,
    stop::Integer = nothing,
)
    head === nothing || start === nothing ||
        throw(ArgumentError("`head` and `start` keywords cannot be used together")
    tail === nothing || stop === nothing ||
        throw(ArgumentError("`tail` and `stop` keywords cannot be used together")
    if head === tail === start === stop === nothing
        head, tail = 0, 1
    else
        if head === nothing && start !== nothing
            head = start - 1
        end
        if tail === nothing && stop !== nothing
            tail = length(str) - stop
        end
    end

    @assert head !== nothing
    @assert tail !== nothing
    # current implementation here
end
```

**Comment by bkamins** [comment]
Agreed, I constantly have a problem with having to write `tail=0` when I need to chop only head.

Implementation needs fixing for cases where `head === start === nothing` as then `  @assert head !== nothing` will fail then I think (similarly for `tail`).

Thank you!

**Comment by Moelf** [comment]
can we actually make `chop(a)` error since the word '[chop](https://www.merriam-webster.com/dictionary/chop)' doesn't imply about which end it's chopping off:
```
transitive verb
1a : to cut into or sever usually by repeated blows of a sharp instrument
b : to cut into pieces —often used with upchop up an onion
c : to weed and thin out (young cotton)
d : to cut as if by chopping chop prices a bridge chops the lake in two
2 : to strike (something, such as a ball) with a short quick downward stroke
3 : to subject to the action of a chopper chop a beam of light
```

**Comment by bkamins** [comment]
this would be breaking

**Comment by Moelf** [comment]
what we're proposing is also technically breaking

**Comment by bkamins** [comment]
ah - yes, you mean @StefanKarpinski proposal.

**Comment by StefanKarpinski** [comment]
Yes, the question is whether anyone is actually using `chop(s, head=h)` to correctly mean `chop(s, head=h, tail=1)` or if everyone who is doing that meant to write the latter and just didn't realize that it didn't mean that already.

**Comment by bkamins** [comment]
I agree.

`chop(x)` has a meaning in other programming languages meaning removing from the right (originally in Perl AFAICT).

Now the difference is that `chop(x)` is probably present in user code much more often than bare `chop(s, head=h)`, as it is extremely unlikely that someone wants to chop `h` characters in the front and `1` character in the end and not write `tail=1` just for clarity as @StefanKarpinski noted above.

Still the problem is that it is breaking (but I would go for it).

**Comment by bkamins** [comment]
@StefanKarpinski - the question is which release of Julia would this go into if we agreed to make this change?

**Comment by StefanKarpinski** [comment]
I'm not suggesting changing the behavior of `chop(s)` — that should continue to do what it does now.

**Comment by bkamins** [comment]
I am sorry if I was not precise. My thinking is that we should leave `chop(x)` as is and for `chop(x, head=h)` to assume that `tail=0`. So I assume we are in agreement here.
